src/pipeline/pipeline_construction.c

```diff
@@ -196,30 +252,95 @@ static ResultProcessor *getArrangeRP(Pipeline *pipeline, const AggregationPipeli
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.1.1 IsInternal");
+          // In non-optimized aggregate queries, we sort maxResultsLimit results
+          rp = RPSorter_NewByFields(params->maxResultsLimit, sortkeys, nkeys, astp->sortAscMap);
+          up = pushRP(&pipeline->qctx, rp, up);
+        } else {
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.1.2 !IsInternal");
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.1.2 maxResults = %zu HasSortBy = %d", maxResults, !!HasSortBy(&params->common));
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.1.2 astp->offset = %zu, astp->limit = %zu", astp->offset, astp->limit);
+          rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
+          up = pushRP(&pipeline->qctx, rp, up);
+        }
+      } else {
+        RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.2 !IsAggregate || !HasDepleter");
+        if (IsAggregate(&params->common) && HasWithCount(&params->common) && HasSortBy(&params->common) && !IsInternal(&params->common)) {
+          // FT.AGGREGATE + WITHCOUNT + SORTBY (no limit, no depleter) -> Add default limit at coordinator
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.2.1 IsAggregate && !HasDepleter");
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.2.1 astp->offset = %zu, astp->limit = %zu", astp->offset, astp->limit);
+          // Support SORTBY + MAX
+          size_t sort_limit = astp->limit ? astp->limit : DEFAULT_LIMIT;
+          rp = RPSorter_NewByFields(sort_limit, sortkeys, nkeys, astp->sortAscMap);
+          up = pushRP(&pipeline->qctx, rp, up);
+        } else {
+          RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.2.2 !IsAggregate || !HasDepleter");
+          rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
+          up = pushRP(&pipeline->qctx, rp, up);
+        }
+      }
     } else if (IsHybrid(&params->common) ||
-               IsSearch(&params->common) && !IsOptimized(&params->common) ||
+               (IsSearch(&params->common) && !IsOptimized(&params->common)) ||
                HasScorer(&params->common)) {
+      RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.3");
       // No sort? then it must be sort by score, which is the default.
       // In optimize mode, add sorter for queries with a scorer.
       rp = RPSorter_NewByScore(maxResults);
       up = pushRP(&pipeline->qctx, rp, up);
+    } else if (IsAggregate(&params->common) && HasDepleter(&params->common)) {
+      RedisModule_Log(RSDummyContext, "notice", "Nafraf: getArrangeRP:2.4 IsAggregate && HasDepleter");
+      // In non-optimized aggregate queries, we need to add a synchronous depleter
+      // Use RPSyncDepleter_New to run synchronously (no background thread)
+      rp = RPSyncDepleter_New();
```

Comment:
Can confirm it is always added when there is no sorter/grouper, which is already depleting?